server/server.go

```diff
@@ -644,7 +646,9 @@ func (s *Server) spawn(l net.Listener, c net.Conn) {
 		tconn.Close()
 		return
 	}
-	limited, err := s.config.isLimited(tconn.ConnectionState())
+	connState := tconn.ConnectionState()
+	certmetrics.Observe(connState.PeerCertificates...)
```

Comment:
not sure if I also need to observe certificates in Verified Chains of ConnectionState here

Reply:
It's good to include the full chain here 👍

Reply:
My apologies, I misunderstood your comment. I thought you were asking whether we needed the full chain from `PeerCertificates` or just the leaf cert (i.e. `PeerCertificates...` vs. `PeerCertificates[0]`), and my comment was supporting the use of `PeerCertificates...`.

I think in practice `PeerCertificates` will be the same as `VerifiedChains` (but I think verified chains will also include the root), so I think it would be best to just observe `PeerCertificates...` here. Worth checking with @maxnystrom.

Reply:
makes sense, ideally `PeerCertificates` chain should be one of the chains in `VerifiedChains` so should just observe certs sent by peer. Also, wondering why root wouldnt be a part of `PeerCertificates` or why we should not include root for observation

Reply:
> Also, wondering why root wouldnt be a part of PeerCertificates 

When a TLS handshake is initiated, the server sends down its leaf certificate, along with any intermediates. The client then checks if it can build a chain to that leaf certificate given the intermediates and the roots it already has embedded in its trust store. The server doesn't include the CA certificate b/c the client must already have it for the handshake to succeed.

With client auth, it's basically the reverse. The server must already have the root CA certificate(s) required to build a chain of trust to the client leaf certificate, or the handshake will fail. Hence I expect `PeerCertificates` to not include the root (though I think it's valid to include it, it would just be ignored by the server).

> or why we should not include root for observation

I didn't mean to imply it would be a problem to include it here, I was just stating my understanding of how these variables differ. Including it won't be a problem, though we know from above it should have already been observed when the server configured the root CAs it allows for client auth.

`VerifiedChains` may also only contain a subset of the certificates from `PeerCertificates`, if the client is sending intermediates the server doesn't need to build the chain of trust. You could argue we only need to report on the relevant certificates, but I think it's best to just report everything the client is sending in `PeerCertificates`.

Reply:
@cbroglie is correct, though https://tools.ietf.org/html/rfc5246#section-7.4.6 indicates that the client certificate is singular, so the client must only send the leaf (end-entity). 

Though, like server certificates, a certificate chain is only valid if all the certificates in the chain are valid. And I have low confidence that all clients only send the leaf so checking `VerifiedChains` makes sense.